app/services/affiliate-slots.js

```diff
@@ -81,133 +89,135 @@ const checkMobileSystem = (unit) => {
-   *
-   * @param {Function[]} filter functions for targeting
-   * @returns {AffiliateUnit[]}
+   * @returns {Targeting[]}
    */
-  getAllUnits(filters = []) {
-    let activeTargeting = targeting
-      // filter targeting by GEO cookie (country)
-      .filter(t => checkFilter(t.country, this.currentCountry))
+  _getTargetingOnSearch(query) {
+    return searchPageTargeting
       // filter targeting by vertical name
       .filter(t => checkFilter(t.vertical, this.currentVertical))
       // filter targeting by current wiki ID
-      .filter(t => checkFilter(t.wikiId, this.currentWikiId));
+      .filter(t => checkFilter(t.wikiId, this.currentWikiId))
+      // filter targeting for search `query`
+      .filter(t => checkFilter(t.query, query));
+  },
 
-    // apply additional criteria
-    filters.forEach((f) => {
-      activeTargeting = activeTargeting.filter(f);
+  fetchUnitForSearch(query, isBig = false) {
+    return new Promise((resolve, reject) => {
+      // check if we have possible units (we can fail early if we don't)
+      if (!this._getAvailableUnits()) {
+        resolve(undefined);
+      }
+
+      // get the units that fulfill the targeting on search
+      const units = _getUnitsForCampaigns(this._getTargetingOnSearch(query))
+        // filter type of ad
+        .filter(u => !!u.isBig === !!isBig)
+        // filter units disabled on search page
+        .filter(u => !u.disableOnSearch)
+      
+      // fetch only the first unit if available
+      const unit = units.length > 0 ? units[0] : undefined
+
+      resolve(unit);
     });
+  },
 
-    // extract unit names from available targeting
-    const availableUnitNames = activeTargeting
-      // grab only names of allowed units
-      .map(t => t.unit)
-      // flatten the array
-      .reduce((prev, curr) => prev.concat(curr), [])
-      // make the list unique
-      .filter(distinct);
-
-    // get all the units based on name and additional checks
-    const availableUnits = units
-      // filter only units that are actually allowed (because of targeting)
-      .filter(u => availableUnitNames.indexOf(u.name) > -1)
-      // check for mobile-system-specific units
-      .filter(u => checkMobileSystem(u))
-      // sort them according to the priority
-      .sort((a, b) => ((a.priority > b.priority) ? 1 : -1));
-
-    return availableUnits;
+  fetchUnitForPage(id, isBig = false) {
+    return new Promise((resolve, reject) => {
+      // check if we have possible units (we can fail early if we don't)
+      if (!this._getAvailableUnits()) {
+        resolve(undefined);
+      }
+      
+      const url = this.fetch.getServiceUrl('taxonomy', `/affiliates/${this.currentWikiId}/${id}`);
+
+      this.fetch.fetchAndParseResponse(url, {}, AffiiatesFetchError)
+        .then((response) => {
+          // convert API to nicer, more useful format
+          const targeting = [];
+
+          response.forEach(e => {
```

Comment:
Any reason to not use `map` here as you need to return an array?

Reply:
it's nested :( I would need to also reduce/flatten it